esphome/components/esp32/__init__.py

```diff
@@ -854,6 +854,10 @@ def _configure_lwip_max_sockets(conf: dict) -> None:
 async def to_code(config):
     cg.add_platformio_option("board", config[CONF_BOARD])
     cg.add_platformio_option("board_upload.flash_size", config[CONF_FLASH_SIZE])
+    cg.add_platformio_option(
+        "board_upload.maximum_size",
+        int(config[CONF_FLASH_SIZE].rstrip("MB")) * 1024 * 1024,
```

Comment:
```suggestion
        int(config[CONF_FLASH_SIZE].removesuffix("MB")) * 1024 * 1024,
```

`rstrip` works and deletes all `M` and all `B` but `removesuffix` is a bit more explicit on the intent